### PR TITLE
Cyborg tablet software system

### PR DESCRIPTION
**Pull request by twilightwanderer**
## About The Pull Request

Adds a system for installing programs to cyborgs depending on their species.

## Why It's Good For The Game

It so happens that in the architecture of the tgstation build, every cyborg owns a tablet. In this form, these tablets are useless, because they are designed only for one program - **robotact**. At the same time, cyborgs have access to install and run any program on any station computer. And instead of using their tablet and station network, the cyborgs have to drive to the computers every time, which seems a little illogical.

I propose to use a system in which each cyborg is recorded a specific set of programs for remote use depending on its specialization.
For cyborgs at the moment I have made an installation of the following programs:

image

**Enginer** - _power_monitor, alarm_monitor, supermatter_monitor_;
**Medical** - _lifeline_;
**Security** - _secureye_;
**Service** - _chatclient_;
**Peacekeeper** - _lifeline, secureye_;
**Clown** - _chatclient_. (HONK in chat)

For the programs **secureye** and **power_monitor** I have created their variants for cyborgs, which are not available for installation on regular tablets and need a network. The regular versions of secureye and power_monitor are not available for installation on tablets.
This set of programs can easily be changed or added to in the future. I plan to do this if my changes are included in the main repository, porting variants of the conventional console interfaces for use in cyborg tablets. 

## Changelog
:cl:
add: Added system for loading programs to cyborgs when selecting a specific model
add: Added two special programs for cyborgs: power_monitor/integrated and secureye/integrated
code: The following files have been changed: robot.dm, tablet_presets.dm, powermonitor.dm, secureye.dm.
/:cl:

### CODE REVIEW

**Comment by FlamingLily** [comment]
Absolutely awesome PR, unfortunately https://github.com/Skyrat-SS13/Skyrat-tg/pull/7367

File: code/modules/modular_computers/computers/item/tablet_presets.dm [hunk 1]
```diff
@@ -110,3 +110,39 @@
...
+	hard_drive.store_file(new /datum/computer_file/program/radar/lifeline)
+
+/obj/item/modular_computer/tablet/integrated/security/Initialize()
+	. = ..()
+	var/obj/item/computer_hardware/hard_drive/small/hard_drive = find_hardware_by_name("solid state drive")
+	hard_drive.store_file(new /datum/computer_file/program/secureye/integrated)
+	
+/obj/item/modular_computer/tablet/integrated/service/Initialize()
+	. = ..()
+	var/obj/item/computer_hardware/hard_drive/small/hard_drive = find_hardware_by_name("solid state drive"
```
Comment (ZephyrTFA) [inline code/modules/modular_computers/computers/item/tablet_presets.dm]:
```suggestion
	var/obj/item/computer_hardware/hard_drive/small/hard_drive = find_hardware_by_name("solid state drive")
```

**Comment by ZephyrTFA** [comment]
Instead of giving them the applications from the installing of their module; give them access to the program downloader. Its much simpler and will be 100% modular

**Comment by twilightwanderer** [comment]
> Instead of giving them the applications from the installing of their module; give them access to the program downloader. Its much simpler and will be 100% modular

Thank you for your response. That's what I thought too at first. It would have been the easiest option. But it looks like the program downloader is unavailable for a reason. The fact is that some programs simply can not be installed on the tablet, other programs are simply unnecessary, and the rest have a modifier _available_on_ntnet = FALSE_.

**Comment by Vizoplez** [comment]
> > Instead of giving them the applications from the installing of their module; give them access to the program downloader. Its much simpler and will be 100% modular
> 
> Thank you for your response. That's what I thought too at first. It would have been the easiest option. But it looks like the program downloader is unavailable for a reason. The fact is that some programs simply can not be installed on the tablet, other programs are simply unnecessary, and the rest have a modifier _available_on_ntnet = FALSE_.

This is already possible, this PR isn't needed!

Cyborgs can have datadisks inserted into them with an open panel, which can contain program data. Furthermore, their tablets themselves are actually incompatible with most of the programs listed here, the mainly functional ones are.

NT CIMS (Crystal monitoring)
SECUREYE
Chat Clients
Outbomb Cuban Pete (don't ask)

while the rest are deprecated.

**Comment by Stalkeros2** [comment]
No.

**Comment by FlamingLily** [comment]
> > Instead of giving them the applications from the installing of their module; give them access to the program downloader. Its much simpler and will be 100% modular
> 
> Thank you for your response. That's what I thought too at first. It would have been the easiest option. But it looks like the program downloader is unavailable for a reason. The fact is that some programs simply can not be installed on the tablet, other programs are simply unnecessary, and the rest have a modifier _available_on_ntnet = FALSE_.

Another problem with doing that is that borgs have all access, so it would make it very easy to get access locked programs, I much prefer this method

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -83,8 +83,31 @@
...
 /mob/living/silicon/robot/proc/create_modularInterface()
-	if(!modularInterface)
+	// TG ORIGINAL BEGIN
+	//if(!modularInterface)
+	//	modularInterface = new /obj/item/modular_computer/tablet/integrated(src)
+	// TG ORIGINAL END
+	
+	//SKYRAT EDIT ADDITION BEGIN
+	//Creating a certain cyborg tablet depending on his species
+	if (model.name == "Default")
```
Comment (FlamingLily) [inline code/modules/mob/living/silicon/robot/robot.dm:95]: This is causing runtimes, model is a null at this point, which makes me theorise that this is called before setting up the model for the borg

**Comment by FlamingLily** [comment]
Infact, thinking about it, this is not a great way to do it. You're effectively changing the internal tablet item whenever there's a model change, which is not a great way to do it. They way I'd do it is

When a cyborg chooses a model, it installs the programs of that model. When it gets a reset, all programs (apart from lawtact or robotact or whatev) get deletes

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 2 above)
Comment (twilightwanderer) [reply code/modules/mob/living/silicon/robot/robot.dm:95]: Okay. I will do a separate procedure for this purpose.

**Comment by twilightwanderer** [comment]
> Infact, thinking about it, this is not a great way to do it. You're effectively changing the internal tablet item whenever there's a model change, which is not a great way to do it. They way I'd do it is
> 
> When a cyborg chooses a model, it installs the programs of that model. When it gets a reset, all programs (apart from lawtact or robotact or whatev) get deletes

I will think about how this can be implemented and I will try to present an implementation according to your advice in the near future

**Comment by twilightwanderer** [comment]
> 
> 
> Infact, thinking about it, this is not a great way to do it. You're effectively changing the internal tablet item whenever there's a model change, which is not a great way to do it. They way I'd do it is
> 
> When a cyborg chooses a model, it installs the programs of that model. When it gets a reset, all programs (apart from lawtact or robotact or whatev) get deletes

I redid it as you recommended.

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 2 above)
Comment (twilightwanderer) [reply code/modules/mob/living/silicon/robot/robot.dm:95]: I restored the standard tablet creation and added a procedure for installing programs

**Comment by jjpark-kb** [comment]
due to being stale and the feature freeze, im going to close this for now. Please reopen this when the feature freeze ends!

**Pull request closed without merging**